### PR TITLE
[docs][example] VLM Examples

### PR DESCRIPTION
**Pull request by nithinvc**
## Summary

Adds two end-to-end multi-turn VLM RL examples (Geometry-3K and VisGym) along with a Vision-Language RL tutorial that documents the shared VLM setup (flags, dataset record shape, local vLLM override). Also wires the `SkyRLVLMGymGenerator` from #1486 into the main entrypoint behind a config flag so VLM runs can be launched end-to-end from `ppo_base_config.yaml`.

- **Geometry-3K example** (`examples/train/geometry3k/`) — multi-turn GRPO on [hiyouga/geometry3k](https://huggingface.co/datasets/hiyouga/geometry3k) with `Qwen/Qwen3-VL-8B-Instruct`. Up to 3 turns per episode; model checks candidate answers with a `calc_score` tool before committing to a final `\boxed{}` answer. Binary reward.
- **VisGym example** (`examples/train/visgym/`) — multi-image multi-turn RL where every env step returns a new image observation. Two recipes:
  - `run_visgym_from_instruct.sh` — vanilla `Qwen3-VL-8B-Instruct`, keyword actions, task-only reward, KL on.
  - `run_visgym_from_sft.sh` — starts from a structured `<observation>/<justification>/<action>` SFT checkpoint with tuple actions and a mixed task+format reward.
- **Docs** — new `tutorials/vision_language_rl.mdx` (shared VLM setup, required flags, dataset shape, support matrix) and example pages for each recipe under `examples/`. Docs pages include reward curves and a VisGym rollout GIF.
- **Generator wiring** — new `generator.vision_language_generator` config flag. When true, `main_base.py` constructs `SkyRLVLMGymGenerator` instead of `SkyRLGymGenerator`. Defaults to false, no behavior change for existing runs.
- **`mm_token_type_ids` shim** (`model_wrapper.py`) — transformers v5 expects `mm_token_type_ids` to be populated at tokenization to distinguish text vs. multimodal tokens, but vLLM doesn't support transformers v5 yet and doesn't return them. Populate here from `image_token_id` when images are present and the field is missing. Remove once vLLM ships transformers v5 support.

## Test plan

- [x] `bash examples/train/geometry3k/run_geometry3k.sh` trains end-to-end on 1×8×H100; reward curve matches `docs/public/images/examples/geometry3k_reward.png`.
- [x] `bash examples/train/visgym/run_visgym_from_instruct.sh` trains end-to-end on 1×8×H100; reward curve matches `docs/public/images/examples/visgym_maze2d_reward.png`.
- [x] `MODEL_PATH=/path/to/sft_ckpt bash examples/train/visgym/run_visgym_from_sft.sh` trains end-to-end.
- [x] Existing non-VLM runs unaffected (`vision_language_generator: false` is the default).
- [x] Docs build: `cd docs && npm run build`.

### CODE REVIEW

File: docs/content/docs/examples/geometry3k.mdx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+If the tool call scores 1.0, the env returns "you can now provide the final answer." If 0.0 and not on the final turn, the env nudges the model to reason differently. If no tool call is found, the env falls back to extracting a boxed answer directly from the text and terminates the episode.
+
+Observations returned by `step()` are plain-text feedback strings (no images) — the model sees the diagram once in the initial prompt and reasons about it across turns without new image inputs. This is the **text-only observation** pattern described in the VLM tutorial.
+
+## Training Configuration
+
+Key VLM-specific overrides from `examples/train/geometry3k/run_geometry3k.sh`:
+
+```bash title="examples/train/geometry3k/run_geometry3k.sh"
+_SKYRL_USE_NEW_INFERENCE=1 uv run --isolated --extra fsdp \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply docs/content/docs/examples/geometry3k.mdx:61]: this is incorrect, it is supported. Two pieces of evidence:
1. This is how I ran the recipe
2. This is the format our gsm8k example follows: https://github.com/NovaSky-AI/SkyRL/blob/main/examples/train/gsm8k/run_gsm8k.sh#L19

File: docs/content/docs/tutorials/vision_language_rl.mdx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+<Callout type="warn">
+**Local vLLM source override required (temporary).** VLM training needs a newer vLLM than the `vllm==0.19.0` pinned in the root `pyproject.toml`. Until the next vLLM release ships with the multimodal rendering support used by SkyRL's new inference stack, clone vLLM locally and point uv at it by adding one line under `[tool.uv.sources]` in the repo root `pyproject.toml`:
+
+```toml
+[tool.uv.sources]
+# ...existing entries (skyrl-gym, torch, etc.)...
+vllm = { path = "/abs/path/to/vllm" }
+```
+
+Use an absolute path. After editing, the standard `uv run --isolated --extra fsdp ...` invocation used by the VLM example scripts will pick up your local vLLM checkout. Remove this override once the next upstream vLLM release is available. Please also unpin the vLLM version:
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply docs/content/docs/tutorials/vision_language_rl.mdx:19]: Isolated does support extra and custom sources. This is how I ran training runs + is consistent with our other examples

File: docs/content/docs/examples/visgym.mdx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+  generator.inference_engine.async_engine=true \
+  generator.sampling_params.temperature=0.7 \
+  # Algorithm — task+format reward, no KL
+  trainer.algorithm.advantage_estimator="grpo" \
+  trainer.algorithm.use_kl_loss=false \
+  environment.env_class=visgym \
+```
+
+```bash title="examples/train/visgym/run_visgym_from_instruct.sh (excerpt)"
+_SKYRL_USE_NEW_INFERENCE=1 uv run --isolated --extra fsdp \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply docs/content/docs/examples/visgym.mdx:148]: see geometry docs example

File: docs/content/docs/tinker/cookbook.mdx [hunk 4]
```diff
@@ -108,3 +108,33 @@ TINKER_API_KEY=tml-dummy uv run --with tinker --with datasets --with torch \
...
+Trains an image classifier on top of [Qwen3-VL](https://huggingface.co/Qwen/Qwen3-VL-8B-Instruct) using the [`vlm_classifier` cookbook recipe](https://github.com/thinking-machines-lab/tinker-cookbook/tree/main/tinker_cookbook/recipes/vlm_classifier). For background on SkyRL's VLM path — dataset record shape, interaction patterns, and the broader support — see the [Vision-Language RL tutorial](../tutorials/vision_language_rl).
+
+<Callout type="warn">
+VLM training requires a newer vLLM than the version pinned in the repo. Clone vLLM from GitHub and point `[tool.uv.sources]` in the root `pyproject.toml` at your local checkout before running. See the [vLLM source override callout](../tutorials/vision_language_rl#requirements) on the tutorial page for the exact steps.
+</Callout>
+
+This recipe uses a **different server command** than the common one at the top of this page, in particular the `use_sample_packing` override is VLM-specific:
+
+```bash
+uv run --isolated --extra tinker --extra fsdp \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply docs/content/docs/tinker/cookbook.mdx:123]: see geometry docs example

File: docs/content/docs/examples/visgym.mdx [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+```
+
+Each row carries `env_class=visgym`, `visgym_env_id`, and optionally a deterministic `seed` for evaluation.
+
+## Training Configuration
+
+Key overrides from both scripts:
+
+```bash title="examples/train/visgym/run_visgym_from_sft.sh (excerpt)"
+_SKYRL_USE_NEW_INFERENCE=1 uv run --isolated --extra fsdp \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply docs/content/docs/examples/visgym.mdx:128]: see geometry docs example

File: examples/train/geometry3k/run_geometry3k.sh [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+: "${NUM_GPUS:=8}"
+
+if [ ! -f "$DATA_DIR/train.parquet" ]; then
+  echo "=== Generating Geometry-3K dataset ==="
+  uv run examples/train/geometry3k/geometry_3k_dataset.py --output_dir "$DATA_DIR"
+fi
+: "${LOGGER:=console}"
+: "${EXPORT_PATH:="$HOME/exports/geometry3k_vlm"}"
+
+_SKYRL_USE_NEW_INFERENCE=1 uv run --isolated --extra fsdp --with pylatexenc \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply examples/train/geometry3k/run_geometry3k.sh:22]: see geometry docs example

File: examples/train/visgym/run_visgym_from_instruct.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+if [ ! -f "$EVAL_DATA_DIR/train.parquet" ]; then
+  echo "=== Generating eval stub dataset for $ENV_ID ==="
+  uv run examples/train/visgym/dataset.py \
+    --env_id "$ENV_ID" \
+    --num_rows 64 \
+    --seed \
+    --output_dir "$EVAL_DATA_DIR"
+fi
+
+_SKYRL_USE_NEW_INFERENCE=1 uv run --isolated --extra fsdp \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply examples/train/visgym/run_visgym_from_instruct.sh:48]: see geometry docs example

File: examples/train/visgym/run_visgym_from_sft.sh [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+if [ ! -f "$EVAL_DATA_DIR/train.parquet" ]; then
+  echo "=== Generating eval stub dataset for $ENV_ID ==="
+  uv run examples/train/visgym/dataset.py \
+    --env_id "$ENV_ID" \
+    --num_rows 64 \
+    --seed \
+    --output_dir "$EVAL_DATA_DIR"
+fi
+
+_SKYRL_USE_NEW_INFERENCE=1 uv run --isolated --extra fsdp \
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply examples/train/visgym/run_visgym_from_sft.sh:46]: see geometry docs example

File: examples/train/geometry3k/math_utils.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+    return True
+
+
+def are_equal_under_sympy(ground_truth_normalized: str, given_normalized: str):
+    are_equal = False
+    try:
+        expr = f"({ground_truth_normalized})-({given_normalized})"
+        if should_allow_eval(expr):
+            sympy_diff = _sympy_parse(expr)
+            simplified = sympy.simplify(sympy_diff)
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply examples/train/geometry3k/math_utils.py:366]: this is okay for an example

File: examples/train/geometry3k/env.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+    def _extract_answer_from_text(self, text: str) -> str | None:
+        """Extract answer: prefer \\boxed{}, fall back to last non-empty line."""
+        boxed = extract_boxed_answer(text)
+        if boxed:
+            return str(boxed).strip()
+        for line in reversed(text.splitlines()):
+            cleaned = line.strip()
+            if cleaned:
+                return cleaned[:512]
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply examples/train/geometry3k/env.py:108]: The method gets 512 characters of the last line, in the above loop `reversed(text.splitlines())`

File: skyrl/train/config/ppo_base_config.yaml
Comment (SumanthRH) [inline skyrl/train/config/ppo_base_config.yaml:1]: The YAML is legacy and will be deleted soon. Revert?

File: examples/train/geometry3k/geometry_3k_dataset.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    # Save the full dataset
+    output_dir = args.output_dir
+    os.makedirs(output_dir, exist_ok=True)
+
+    train_parquet_path = os.path.join(output_dir, "train.parquet")
+    train_dataset.to_parquet(train_parquet_path)
+    print(f"Saved full training set ({len(train_dataset)} examples) to {train_parquet_path}")
+
+    # Process and save the val split
+    if "val" in dataset:
```
Comment (SumanthRH) [inline examples/train/geometry3k/geometry_3k_dataset.py]: The split name should be `validation` according to https://huggingface.co/datasets/hiyouga/geometry3k/viewer/default/validation, not `val`

File: examples/train/geometry3k/geometry_3k_dataset.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        val_dataset = dataset["val"]
```
Comment (SumanthRH) [inline examples/train/geometry3k/geometry_3k_dataset.py]: The split name is `validation` not `val` according to https://huggingface.co/datasets/hiyouga/geometry3k/viewer/default/validation

File: examples/train/geometry3k/math_utils.py [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+
+
+def extract_answer(passage: str) -> str | None:
+    """Extract a boxed answer from the passage, or return None."""
+    if "\\boxed" in passage:
+        return extract_boxed_answer(passage)
+    return None
+
+
+def grade_answer_verl(solution_str: str, ground_truth: str) -> bool:
```
Comment (SumanthRH) [inline examples/train/geometry3k/math_utils.py]: Let's instead call this `grade_answer_from_boxed` and add the reference used (looks like it's VERL)  as a comment

File: docs/content/docs/examples/geometry3k.mdx [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+  generator.vision_language_generator=true \
+  generator.batched=false \
+  trainer.use_sample_packing=false \
+  # Multi-turn rollouts
+  generator.max_turns=3 \
+  generator.inference_engine.async_engine=true \
+  # Algorithm
+  trainer.algorithm.advantage_estimator="grpo" \
+  trainer.algorithm.use_kl_loss=false \
+  generator.n_samples_per_prompt=8 \
```
Comment (SumanthRH) [inline docs/content/docs/examples/geometry3k.mdx]: This snippet here says `n_samples_per_prompt=8` but then the actual script at  `examples/train/geometry3k/run_geometry3k.sh` says `n_samples_per_prompt=4`.

File: docs/content/docs/tutorials/vision_language_rl.mdx [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+---
+
+
+SkyRL supports multi-turn RL training on vision-language models (VLMs) — the policy reads image observations, reasons, and emits actions or tool calls, with images flowing through the conversation across turns. This guide covers the setup common to all VLM training runs in SkyRL: enabling the VLM code path, the required flags, the dataset record shape, and the current support matrix. For concrete runnable recipes, see the [Geometry-3K](../examples/geometry3k) and [VisGym](../examples/visgym) example pages.
+
+## Requirements
+
+<Callout type="warn">
+**Local vLLM source override required (temporary).** VLM training needs a newer vLLM than the `vllm==0.19.0` pinned in the root `pyproject.toml`. Until the next vLLM release ships with the multimodal rendering support used by SkyRL's new inference stack, clone vLLM locally and point uv at it by adding one line under `[tool.uv.sources]` in the repo root `pyproject.toml`:
+
```
Comment (SumanthRH) [inline docs/content/docs/tutorials/vision_language_rl.mdx:12]: Can you specify the exact commit required? It looks like it needs to be after [80b1823](https://github.com/vllm-project/vllm/commit/80b18230e0ffe94f8198862675a0dc5e86c10997)

**Review by SumanthRH — COMMENTED** [review]
Overall LGTM,  next time around let's break such changes up into smaller PRs. Have you run the dataset preparation scripts E2E? let's make sure we do that.

File: docs/content/docs/tutorials/vision_language_rl.mdx (hunk 15 above)
Comment (nithinvc) [reply docs/content/docs/tutorials/vision_language_rl.mdx:12]: done

File: docs/content/docs/examples/geometry3k.mdx (hunk 14 above)
Comment (nithinvc) [reply docs/content/docs/examples/geometry3k.mdx]: fixed, both are now 4 samples per prompt

File: examples/train/geometry3k/math_utils.py (hunk 13 above)
Comment (nithinvc) [reply examples/train/geometry3k/math_utils.py]: renamed

File: skyrl/train/config/ppo_base_config.yaml
Comment (nithinvc) [reply skyrl/train/config/ppo_base_config.yaml:1]: reverted

File: examples/train/geometry3k/geometry_3k_dataset.py (hunk 11 above)
Comment (nithinvc) [reply examples/train/geometry3k/geometry_3k_dataset.py]: fixed

File: examples/train/geometry3k/geometry_3k_dataset.py (hunk 12 above)
Comment (nithinvc) [reply examples/train/geometry3k/geometry_3k_dataset.py]: fixed

**Comment by nithinvc** [comment]
> Overall LGTM, next time around let's break such changes up into smaller PRs. Have you run the dataset preparation scripts E2E? let's make sure we do that.

Yes, the "validation" key got past me since the script training run uses the test split `test.parquet` when doing evals. Just reran both dataset generation scripts and they produce the correct parquet files

**Review by SumanthRH — APPROVED** [review]
